Make the code PEP8 compliant.

jensrudolf opened the pull request:
Hi @noseka1,

I really like your project Linuxband and do want contribute at little to keep it alive.

This commit tries to make the code more [PEP8](https://www.python.org/dev/peps/pep-0008/) compliant by fixing most of
the reports generated by the [flake8](https://gitlab.com/pycqa/flake8) tool. Contrary to PEP8, I set the
maximum line length to a more reasonable value of 120 characters.

There are some whitespace reportings left, especially regarding the file
chord_table.py. However, this is intended to keep the chords tables
readable and totally consistent with [PEP20](https://www.python.org/dev/peps/pep-0020/) (The Zen of Python).

This work is preparing a possible port to Python 3 as Python 2 is close to [EOL.](https://pythonclock.org/)

The following changes have been made:

 - Cleanup unnecessary blank lines
 - Add blank lines where necessary
 - Use 'is' and  'is not' operator for comparisons to None
 - Split the use of multiple statements on one line (E701)
 - Remove semicola from statements (E703)
 - Remove unnecessary whitespaces (E201, E202)
 - Add whitespaces where necessary
 - Remove some useless inline comments
 - Replace '<>' with '!=' (W603)
 - Remove redundant backslash (E502)
 - Remove mixed indention with tabstops and whitespace
 - Sort module imports according to PEP8
 - Remove line break before binary operator (W503)
 - Fix ambigous variable naming (E741)
 - Fix usage of bare 'except' (E722)
 - Replace escape sequences with corresponding utf-8 symbol (W605)
 - Improve indention and alignment to increase readability
 - Fix wrong comment indention
 - Align line continuations with visual indent
 - Reduce length of some lines to keep em < 120 chars

Feedback is greatly appreciated.

noseka1 commented on src/main/python/linuxband/config.py:
```diff
@@ -127,7 +130,7 @@ def __load_default_config(self):
             self.__config.read(Config.__default_config)
             self.set_work_dir(Config.__home_dir)
-        except:
+        except IOError:
```
Same comment as before.

jensrudolf replied:
Again `ConfigParser.Error`

noseka1 commented on src/main/python/linuxband/gui/chord_entries.py:
```diff
@@ -164,7 +171,7 @@ def __get_chord_names(self):
         chord_names = []
         for c in base_ext:
-            for k, v in chordlist.iteritems(): #@UnusedVariable
+            for k, v in chordlist.iteritems():
```
I used the #@UnusedVariable to suppress a warning in Eclipse. What about following this advice: https://stackoverflow.com/questions/13944234/how-to-suppress-unused-variable-warnings-in-eclipse-pydev/13944314#13944314

This would work for me:
for k, _ in chordlist.iteritems():

jensrudolf replied:
Do you have any objections against using `cordlist.iterkeys()`? This seems to be available in Python2.

noseka1 replied:
`cordlist.iterkeys()` sounds great!

noseka1 commented on src/main/python/linuxband/gui/events/groove.py:
```diff
@@ -36,34 +39,38 @@ def __init__(self, glade, grooves):
 
     def on_treeview1_cursor_changed_callback(self, treeview):
         """ User clicked on the groove in the first column. """
-        path, column = treeview.get_cursor() #@UnusedVariable
```
Same as before.

jensrudolf replied:
In these cases (Treeview object), I would prefer to change that to `path = treeview.get_cursor()[0]` to avoid the '_' character. How about that?

noseka1 replied:
Yeah, great idea!

noseka1 commented on src/main/python/linuxband/gui/events/groove.py:
```diff
@@ -36,34 +39,38 @@ def __init__(self, glade, grooves):
     def on_treeview2_cursor_changed_callback(self, treeview):
         """ User clicked on the variation of groove. """
-        path, column = self.__treeview1.get_cursor() #@UnusedVariable
+        path, column = self.__treeview1.get_cursor()
```
Same as before.

noseka1 commented on src/main/python/linuxband/gui/events/groove.py:
```diff
@@ -36,34 +39,38 @@ def __init__(self, glade, grooves):
+        path, column = self.__treeview1.get_cursor()
         model = self.__groovesModel[path[0]][6]
-        path, column = self.__treeview2.get_cursor() #@UnusedVariable
+        path, column = self.__treeview2.get_cursor()
```
Same as before.

noseka1 commented on src/main/python/linuxband/gui/events/groove.py:
```diff
@@ -112,7 +119,7 @@ def __update_groove_info(self, gr):
         tb = self.__textbuffer2
         tb.set_text('')
-        start, end = tb.get_bounds() #@UnusedVariable
+        start, end = tb.get_bounds()
```
Same as before.

jensrudolf replied:
Also here, I would prefer to change that to `tb.get_bounds()[0]`.

noseka1 commented on src/main/python/linuxband/gui/gui_logger.py:
```diff
@@ -29,13 +31,16 @@ def __init__(self, textView, textBuffer):
+                tag = 'fg_red'
 
-            start, end = self.__textBuffer.get_bounds() #@UnusedVariable
+            start, end = self.__textBuffer.get_bounds()
```
Same as before.

jensrudolf replied:
Again `get_bounds()[0]`.

noseka1 commented on src/main/python/linuxband/gui/gui_logger.py:
```diff
@@ -50,7 +55,7 @@ def __init__(self, fmt=None, datefmt=None):
 
         def formatException(self, ei):
-            excType, excValue, excTraceback = ei #@UnusedVariable
+            excType, excValue, excTraceback = ei
```
Same as before.

noseka1 commented on src/main/python/linuxband/midi/midi_player.py:
```diff
@@ -71,17 +76,17 @@ def startup(self):
         try:
             self.__player = subprocess.Popen(command, stdin=subprocess.PIPE)
-        except:
+        except subprocess.CalledProcessError:
```
In addition to subprocess.CalledProcessError, subprocess.Popen(...) can throw other types of exceptions. To be on the safe side, I suggest reverting this change.

noseka1 commented on src/main/python/linuxband/midi/midi_player.py:
```diff
@@ -105,15 +110,15 @@ def shutdown(self):
         try:
             os.close(self.__piper)
-        except:
+        except IOError:
```
I think os.close() can throw OSError as well. To remain on the safe side I would better stick with catching all possible exceptions.
Btw. swallowing the exception like this is a bad code. It would have been much better to log the exception.

noseka1 commented on src/main/python/linuxband/midi/midi_player.py:
```diff
@@ -105,15 +110,15 @@ def shutdown(self):
         try:
             os.close(self.__pipew)
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/midi_player.py:
```diff
@@ -105,15 +110,15 @@ def shutdown(self):
         try:
             os.close(self.__pout)
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -33,17 +36,17 @@ def check_mma_syntax(self, mma_data):
         try:
             mma = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
-        except:
+        except subprocess.CalledProcessError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -33,17 +36,17 @@ def check_mma_syntax(self, mma_data):
             fout.write(mma_data)
             fout.close()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -65,11 +68,11 @@ def generate_smf(self, mma_data):
             try:
                 os.close(piper)
-            except:
+            except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -65,11 +68,11 @@ def generate_smf(self, mma_data):
             try:
                 os.close(pipew)
-            except:
+            except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -79,17 +82,17 @@ def __do_generate_smf(self, piper, pipew, mma_data):
         try:
             mma = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
-        except:
+        except subprocess.CalledProcessError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -79,17 +82,17 @@ def __do_generate_smf(self, piper, pipew, mma_data):
             fout.write(mma_data)
             fout.close()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -100,9 +103,9 @@ def __do_generate_smf(self, piper, pipew, mma_data):
+            os.close(pipew)  # close our write pipe end, now only MMA has it opened
             midi_data = fin.read()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/midi/mma2smf.py:
```diff
@@ -111,19 +114,23 @@ def __do_generate_smf(self, piper, pipew, mma_data):
+        """
         Parse out the error line number. Error line example: ERROR:<Line 23><File:/proc/self/fd/0>
         """
-        for i, line in enumerate(lines): #@UnusedVariable
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/mma/grooves.py:
```diff
@@ -95,12 +99,13 @@ def __do_load_grooves(self, grooves_list, path):
         Load grooves and recurse into subdirectories.
         """
-        for dirname, dirnames, filenames in os.walk(path): #@UnusedVariable            
+        for dirname, dirnames, filenames in os.walk(path):
```
Same comment as before.

jensrudolf replied:
Seems to me, that all three variable are used below, hence I think, the Unused directive is not necessary.

noseka1 commented on src/main/python/linuxband/mma/grooves.py:
```diff
@@ -128,14 +133,14 @@ def __parseGrooves(self, file_name):
                 logging.exception("Failed to parse the file.")
             mma_file.close()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/mma/grooves.py:
```diff
@@ -145,7 +150,7 @@ def __load_grooves_from_cache(self):
             finally:
                 infile.close()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/mma/grooves.py:
```diff
@@ -163,5 +168,5 @@ def __cache_grooves(self, grooves_list):
             finally:
                 outfile.close()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/mma/parse.py:
```diff
@@ -142,21 +144,23 @@ def parse(inpath):
         # track function BASS/DRUM/APEGGIO/CHORD ...
         if '-' in action:
-            trk_class, ext = action.split('-', 1) #@UnusedVariable
+            trk_class, ext = action.split('-', 1)
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/mma/song.py:
```diff
@@ -38,7 +42,7 @@ def load_from_file(self, file_name):
             finally:
                 mma_file.close()
-        except:
+        except IOError:
```
Same comment as before.

noseka1 commented on src/main/python/linuxband/mma/song.py:
```diff
@@ -128,5 +135,5 @@ def __do_write_to_file(self, file_name, data):
             finally:
                 f.close()
-        except:
+        except IOError:
```
Same comment as before.

jensrudolf commented on src/main/python/linuxband/config.py:
```diff
@@ -47,7 +50,7 @@ def load_config(self):
                 logging.debug("Opening config")
                 self.__config.read(Config.__rc_file)
-            except:
+            except IOError:
```
OK, got it. But I still like to avoid empty except clauses. For any ConfigParser related call I would like to catch for `ConfigParser.Error`, the base class all exceptions thrown by ConfigParser.

noseka1 replied:
I'm not sure how you want to handle the ConfigParser.Error after you caught it. Could you provide a sample code? And what about other possible exceptions? In the future, if the ConfigParser introduces further exceptions, I would like to make sure we still catch them.

I agree that an empty except: clause is a bad code. How about just logging the exception:
except:
  logging.debug(e)

jensrudolf commented:
I refactored the code to omit the "unused variables" by unpacking the tuples or using the underscore character, where necessary.
Furthermore, I reworked the exception handling to catch the exceptions that can be thrown according to the std lib's documentation. Additionally, the exception message is appended to the logging output in order to  ease debugging. There is still room for improvement. However, porting the code to Python 3 eventually might simplify certain cases (e.g. by using context managers).

What do you think about it?

jensrudolf commented:
Hi @noseka1, did you find some time to look into the changes of my latest commit 5fe2743? What do you think about the exception handling?

jensrudolf commented:
Sorry, the close was unintentionally. I hit the wrong button :-(

noseka1 commented on src/main/python/linuxband/gui/events/groove.py:
```diff
@@ -119,7 +119,7 @@ def __update_groove_info(self, gr):
         tb = self.__textbuffer2
         tb.set_text('')
-        start, end = tb.get_bounds()
+        start, end = tb.get_bounds()[0]
```
end variable should likely be removed here

noseka1 replied:
Actually no, the start variable is the one unused here

noseka1 replied:
I see, it's already fixed in the next commit. Disregard.

noseka1 commented:
I was thinking about the exception handling and reviewed your latest approach. I like that you drop e.strerror to the logs.

I still don't feel comfortable about limiting the scope of caught exceptions, i.e. changing
```
except:
  # all exception handled here
```
to
```
except ConfigParser.Error as e:
  # not really all possible exceptions handled here
```
If it would work for you, I think we could do:
```
except Exception as e:
  # almost all exceptions handled here
```

noseka1 commented:
One more comment regarding the size of this pull request. There is many code changes included on this pull request. These are valuable code changes that I would like to get into the master branch, however, it's difficult to review and merge everything at once.

I suggest to break this pull request up into multiple pull requests. For each item on the "The following changes have been made:" list there could be a separate pull request. That will allow for easier review and also much clearer commit history. What do you think?

jensrudolf commented:
Regarding the exception handling when parsing/writing the config file, I just had a closer look at the Python 2.7 implementation of ConfigParser, especially its read() and write() methods. Judging from the code, I would say the read() method itself raises only exceptions based on ConfigParser.Error. However, the ConfigParser read() and write() methods internally use readline() and write() on File-like objects, which may throw IOError. Hence, I would suggest to catch ConfigParser.Error and IOError. All other Exceptions that could theoretically appear during read() or write() (e.g. KeyboardInterrupt or some runtime error) I would like to catch at a somehow more global point (maybe in Gui()) as they are not really related to the parsing the config file. What do you think?

jensrudolf commented:
I totally agree with your last point, that this merge request is quite big. As you suggest, I will break this up into several smaller requests. I will try to split this up into pieces during the coming days as I find some time. Therefore, I close this one in favor of the split requests. Thanks for your comments.

noseka1 commented:
Thanks for investigating further on the exception topic. I wish the ConfigParser library would specify and limit which exact exceptions can be thrown by its APIs. I understand that you can review the ConfigParser's source code and draw conclusions about the current set of exceptions that can be thrown but what about the future ConfigParser implementations? In the future, ConfigParser can change the implementation and throw additional exceptions. Where I am getting at is that unless there is a clear API contract provided by the ConfigParser library I would prefer the defensive LinuxBand code that won't break in the future because of the possible changes in ConfigParser.

Handling exceptions at a more global point might be technically difficult as the LinuxBand code is pretty much a set of handlers called back by the UI library.

Thanks a lot for breaking up the pull request in smaller pieces. That's great. In the meantime I am working on porting LinuxBand from GTK2 to GTK3. I think that this needs to happen before migrating to Python 3.
